go/ai/retriever.go

```diff
@@ -55,58 +71,154 @@ type RetrieverResponse struct {
-func LookupRetriever(provider, name string) *Retriever {
-	return (*Retriever)(core.LookupActionFor[*RetrieverRequest, *RetrieverResponse, struct{}](atype.Retriever, provider, name))
+func LookupRetriever(provider, name string) Retriever {
+	return (*RetrieverAction)(core.LookupActionFor[*RetrieverRequest, *RetrieverResponse, struct{}](atype.Retriever, provider, name))
 }
 
-// Index runs the given [Indexer].
-func (i *Indexer) Index(ctx context.Context, req *IndexerRequest) error {
+// Index runs the given [IndexerAction].
+func (i *IndexerAction) Index(ctx context.Context, req *IndexerRequest) error {
 	if i == nil {
 		return errors.New("Index called on a nil Indexer; check that all indexers are defined")
 	}
 	_, err := (*indexerAction)(i).Run(ctx, req, nil)
 	return err
 }
 
-// Retrieve runs the given [Retriever].
-func (r *Retriever) Retrieve(ctx context.Context, req *RetrieverRequest) (*RetrieverResponse, error) {
+// Retrieve runs the given [RetrieverAction].
+func (r *RetrieverAction) Retrieve(ctx context.Context, req *RetrieverRequest) (*RetrieverResponse, error) {
 	if r == nil {
 		return nil, errors.New("Retriever called on a nil Retriever; check that all retrievers are defined")
 	}
 	return (*retrieverAction)(r).Run(ctx, req, nil)
 }
 
-func (i *Indexer) Name() string { return (*indexerAction)(i).Name() }
+// retrieveParams represents various params of the Retrieve call.
+type retrieveParams struct {
+	request *RetrieverRequest
+}
+
+// generateOption configures params of the Retrieve call.
+type retrieveOption func(req *retrieveParams) error
+
+// WithRetrieverText adds a simple text as document to RetrieveRequest.
+func WithRetrieverText(prompt string) retrieveOption {
+	return func(req *retrieveParams) error {
+		req.request.Document = DocumentFromText(prompt, nil)
+		return nil
+	}
+}
+
+// WithRetrieverDoc adds a document to RetrieveRequest.
+func WithRetrieverDoc(doc *Document) retrieveOption {
+	return func(req *retrieveParams) error {
+		req.request.Document = doc
+		return nil
+	}
+}
+
+// WithRetrieverOpts retriever options to RetrieveRequest.
+func WithRetrieverOpts(opts any) retrieveOption {
+	return func(req *retrieveParams) error {
+		req.request.Options = opts
+		return nil
+	}
+}
+
+// Retrieve calls the retrivers with provided options.
+func Retrieve(ctx context.Context, r Retriever, opts ...retrieveOption) (*RetrieverResponse, error) {
+	req := &retrieveParams{
+		request: &RetrieverRequest{},
+	}
+	for _, with := range opts {
+		err := with(req)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return r.Retrieve(ctx, req.request)
+}
+
+// indexerParams represents various params of the Index call.
+type indexerParams struct {
+	request *IndexerRequest
+}
+
+// generateOption configures params of the Index call.
+type indexerOption func(req *indexerParams) error
+
+// WithIndexerText adds a simple text as document to IndexRequest.
+func WithIndexerText(prompts ...string) indexerOption {
```

Comment:
Is this useful except for tests and demos? Also, why is the arg called "prompts"?

Reply:
Removed. It's probably only useful in extremely simple cases, but most of the time users would want a full document with metadata.